[ZIP TBD]: Circle STARK Verification as a Transparent Zcash Extension (#1107)
zcash/zips: zips/draft-stark-verify-nu.md

    ZIP: XXX
    Title: Circle STARK Verification as a Transparent Zcash Extension

>     Owners: Abdel <abdel.dev.bitcoin@proton.me>, Michael Zaikin <michael.z@starkware.co>

Reviewer: ```suggestion
    Owners: Abdel <abdel.dev.bitcoin@proton.me>
            Michael Zaikin <michael.z@starkware.co>
```

    Status: Draft
    Category: Consensus
    Created: 2025-10-14
    License: MIT
    Discussions-To: <https://forum.zcashcommunity.com/t/stark-verification-as-a-transparent-zcash-extension-tze-i-e-enabler-for-a-starknet-style-l2-on-zcash/52486>

# Terminology

The key words "MUST", "SHOULD", and "MAY" are to be interpreted as described in BCP 14 [^BCP14] when, and only when, they appear in all capitals.

## Abstract

> This ZIP defines a new Transparent Zcash Extension that verifies a bounded‑size Circle STARK proof (Stwo). The extension has a single `type` and a single mode, and exposes a compact precondition/witness interface suitable for verifying L2 validity proofs on Zcash L1. It specifies:

Reviewer: "L1" and "L2" are not defined (and I don't want readers to rely on intuition from other chains because I think it will be useful to define them using other Zcash terminology).

- If `bind_ctx` is set, `ctx_digest` MUST equal the transaction's non‑malleable digest [^zip-0244]. A mismatch MUST cause rejection.

### Witness (prefix‑free)

```
vec  proof            // the Stwo (Circle STARK) proof object (len + bytes)
```

The `proof` MUST conform to the Stwo wire format pinned by `param_id`.

### Verification

`bool tze_verify(0, precondition, witness, context)`:

1. Parse `param_id`; load the exact pinned parameter set. Reject if unknown.
2. Parse `flags`; if any reserved bits set, reject.
3. If `bind_ctx`, compute the required `ctx_digest` from `context` and compare; on mismatch, reject. ([Zcash Zips][3])
4. Parse `pub_in` and `proof`. Enforce all consensus bounds (see Bounds).
5. Initialize the Stwo transcript/channel, **absorb `pub_in` exactly**, and verify `proof` using the parameters pinned by `param_id`.
6. Return `true` iff the verifier accepts; otherwise return `false`.

### Parameter sets (`param_id`)

The namespace is 16‑bit (0x0001..0xFFFF). This ZIP introduces:

- `param_id = 0x0001` — `STWO_V1_P1`

  - Upstream: `starkware-libs/stwo` at tag TBD, commit `TBD`.

>   - Wire format: Stwo proof object at v1.0.0 (exact byte layout pinned in this ZIP's test‑vector directory).

Reviewer: We don't normally rely on test vectors as part of a specification. Rather, test vectors are supposed to be a redundant check that can be compared against both the specification and an implementation. (Ditto for the four lines below.)

Nodes that do not implement this TZE will reject transactions that use it after activation. Because TZE uses a distinct TZE unspent set and separate fields (`tze_inputs`/`tze_outputs`), existing transparent and shielded logic is unaffected.

## Deployment

- Network upgrade: Activate in a future NU alongside [^zip-0244] and [^zip-0245].
- Activation gating: Require published test vectors and an external review of the Stwo binding.

## Security Considerations

- DoS safety: Strict size caps; streaming, fail‑fast parsing; bounded allocations; early rejection on malformed encodings.
- Determinism & non‑malleability: Encodings are prefix‑free; any malleation of `precondition` MUST cause rejection; `tze_verify` is deterministic per [^zip-0222].
- Hash assumptions: Security reduces primarily to collision/preimage resistance of the pinned hash suite plus IOP soundness. No EC‑based hash is assumed; the specifics are pinned by `param_id`.
- Implementation risk: Treat the Rust verifier as consensus‑critical code; pin commit/tag and compiler settings; include cross‑checks against test vectors.

## Reference Implementation

> TBD

## Test Vectors

> TBD

## References

> - [ZIP‑222: Transparent Zcash Extensions (TZE)](https://zips.z.cash/zip-0222)

Reviewer: Nit: this is the wrong format for Markdown references.

ZIP: XXX
Title: Circle STARK Verification as a Transparent Zcash Extension
Owners: Abdel <abdel.dev.bitcoin@proton.me>
Status: Draft
Category: Consensus
Created: 2025-10-14
License: MIT

> Discussions-To: <https://forum.zcashcommunity.com/t/stark-verification-as-a-transparent-zcash-extension-tze-i-e-enabler-for-a-starknet-style-l2-on-zcash/52486>

Reviewer: ```suggestion
    ZIP: XXX
    Title: Circle STARK Verification as a Transparent Zcash Extension
    Owners: Abdel <abdel.dev.bitcoin@proton.me>
    Status: Draft
    Category: Consensus
    Created: 2025-10-14
    License: MIT
    Discussions-To: <https://forum.zcashcommunity.com/t/stark-verification-as-a-transparent-zcash-extension-tze-i-e-enabler-for-a-starknet-style-l2-on-zcash/52486>
```

# Terminology

> The key words **MUST**, **MUST NOT**, **SHOULD**, **SHOULD NOT**, and **MAY** are to be interpreted as described in BCP 14 (RFC 2119 and RFC 8174).

Reviewer: ```suggestion
The key words "MUST", "SHOULD", and "MAY" are to be interpreted as described in BCP 14 [^BCP14] when, and only when, they appear in all capitals.
```
The convention is for this boilerplate to only mention keywords the document actually uses. Also remove the boldface formatting from keyword uses. (Boldface is used for them in the protocol spec but not in ZIPs.)

Stwo provides a modern, high‑performance implementation of **Circle STARKs** in Rust. Pinning a specific Stwo parameter set behind a small `param_id` allows a **minimal consensus change** with tight bounds.

## Requirements

The Specification **MUST** ensure:

1. A **single** TZE `type` (numeric ID assigned by Editors) is defined for **Stwo/Circle STARK verification**. Transactions using this type **MUST** set `tze_mode = 0`.
2. A deterministic `tze_verify(mode, precondition, witness, context)` with **no implementation‑defined behavior**. 
3. Fully specified, **prefix‑free** encodings of `precondition` and `witness` for this `(type,mode)`. 
4. A `param_id` namespace that **pins**: Stwo release/commit, the proof object wire format, field encodings, transcript/Merkle hash suite and personalization, FRI expansion and query caps, any grinding bounds, and any batch‑verification settings.
5. Strict consensus bounds (`PUBIN_MAX_BYTES`, `PROOF_MAX_BYTES`, and parameter caps).
6. Digest integration that follows ZIP‑244/245 (v5) or ZIP‑246 (v6), such that the **precondition** contributes to effecting digests (txid) and the **witness** does not.

> 7. Reference test vectors (valid and invalid) and a minimal reference verifier wrapper **MUST** be provided before activation.

Reviewer: Point 1 has specification in the Requirements section. Point 7 is valid but it shouldn't use the capitalized MUST. The fact that this section places requirements on the specification is implicit and needn't be stated. Also, requirements should be kept high-level.

Pinning an Stwo release/commit is not consistent with defining the semantics by specification rather than implementation &mdash; what if that commit has bugs? It is fine for this document to link to a reference implementation at a particular commit for each `param_id`, and a Zcash full node implementation **may** choose to actually implement it by using that commit (assuming no bugs have yet been found), but that can't be relied on to pin the specification of behaviour.

```suggestion
* Define a TZE `type` for **Stwo/Circle STARK verification**, and fully specify how this type is to be encoded, verified, and integrated into transaction digest computation, without allowing implementation‑defined behavior.
* Define a `param_id` namespace that pins the specification of the Stwo proof relation, the proof object wire format, field encodings, transcript/Merkle hash suite and personalization, FRI expansion and query caps, any grinding bounds, and any batch‑verification settings.
* A reference implementation of the prover and verifier, and reference test vectors (valid and invalid) must be provided for each such `param_id`.
* Limits on proof sizes, etc. must be chosen to avoid potential denial of service, and the necessary updates to ZIP 317 fee calculation [^zip-0317] must be specified.
```
